### ISSUE TITLE
one-glm 一个更高效的glm实现

### ISSUE THREAD

**Issue by BBuf** [Oneflow-Inc/one-glm#15]
## 背景

目前，存在几种不同的预训练模型架构：仅实现编码器架构的自动编码模型（例如BERT），仅实现解码器的自回归模型（例如GPT），以及同时实现编码器和解码器的编码器-解码器模型（例如T5）。[GLM](https://arxiv.org/abs/2103.10360)模型与这些模型略有不同。它采用了一种自回归的空白填充方法， 并且在NLP领域三种主要的任务(自然语言理解，无条件生成，有条件生成)上都取得了不错的结果。GLM被开源在了：https://github.com/THUDM/GLM 。

GLM的开源代码主要是基于PyTorch，DeepSpeed以及Apex来实现的，并且基于deepspeed提供的数据并行和模型并行技术成功训练了GLM-Large（335M），GLM-515M（515M），GLM-10B（10B）等大模型。

为了探索深度学习框架 OneFlow 在训练 GLM 大模型方面的性能表现，我们将原始的 [GLM](https://github.com/THUDM/GLM) 项目移植到了使用 OneFlow 后端进行训练的项目 [one-glm](https://github.com/Oneflow-Inc/one-glm) 。得益于 oneflow 和 pytorch 良好的兼容性我们快速且平滑的移植了 glm 并成功跑通了预训练任务（训练 GLM-large）。此外，由于 oneflow 已经包含了 deepspeed 和 apex 的很多功能和优化技术，天然支持训练大模型，所以我们做到了零插件训练 GLM 模型。最后，我们针对当前移植的 glm 模型做了一些简单的调优工作，优化使用 oneflow 训练 GLM-large 的性能以及显存占用，并取得了相比于原始仓库 2.5 倍的加速以及接近 2 倍的显存节省。我们接下来介绍一下性能，显存的对比表现以及详细的技术细节。

- one-glm 链接：https://github.com/Oneflow-Inc/one-glm
- oneflow 链接：https://github.com/Oneflow-Inc/oneflow

## 性能和显存表现

下面我们先展示一下分别使用官方的 GLM 仓库以及 one-glm 仓库训练 GLM-large 网络的性能和显存表现（数据并行技术），硬件环境为 A100 PCIE 40G，BatchSize 设置为 32。

![图片](https://user-images.githubusercontent.com/35585791/211490102-dc058e77-b118-432b-88a1-f42140e58f7e.png)

![图片](https://user-images.githubusercontent.com/35585791/211490128-1d90b8ef-f533-4b26-a5fc-69792cc767e6.png)

![图片](https://user-images.githubusercontent.com/35585791/211490160-57e6de3f-dd09-4176-87fd-df818d6ec1b0.png)

![图片](https://user-images.githubusercontent.com/35585791/211490175-2147dae5-204b-43a5-892a-d0414e484435.png)

可以看到在 GLM-large 的训练任务中，OneFlow的性能和显存占用大幅度优于原始的基于 PyTorch，DeepSpeed，Apex 的 GLM 实现。

## 将 GLM 迁移到 one-glm 所做的改动

- 将 `from apex.optimizers import FusedAdam as Adam` 替换为 `from oneflow.optim import Adam`
- 将 `from apex.normalization.fused_layer_norm import FusedLayerNorm as LayerNorm` 替换为 `from oneflow.nn import LayerNorm`
- 注释掉 `torch.distributed.ReduceOp`，`torch.distributed.new_group`, `torch.distributed.TCPStore` 这些api。原因是因为这个几个 api 属于 pytorch ddp的新特性，oneflow没有支持。
...

### loss 计算部分的优化    

在原始的 GLM 实现中，loss计算部分使用到了 `mpu.vocab_parallel_cross_entropy` 这个函数 (https://github.com/THUDM/GLM/blob/main/pretrain_glm.py#L263) ，通过对[这个函数](https://github.com/THUDM/GLM/blob/main/mpu/cross_entropy.py#L25-L104)的分析，我们发现实际上它实际上就是实现了 `sparse_softmax_cross_entropy` 的功能，但是在实现的过程中原始的 GLM 仓库使用了 PyTorch 的 autograd.Function 模块并且使用了大量的小算子来拼接出 `sparse_softmax_cross_entropy` 整体的功能。而在 OneFlow 的算子库中已经有 `sparse_softmax_cross_entropy` 这个算子对应的 cuda 实现了，也就是 `flow.sparse_softmax_cross_entropy` 这个 api。所以我们将 GLM 对 `sparse_softmax_cross_entropy` 的 naive 实现替换为 `flow.sparse_softmax_cross_entropy` 这个api，并进行了loss对齐实验。下面展示了基于 oneflow 的 graph 模式训练 GLM-large 模型前 1000 轮的loss 对齐情况，分别测试了 FP32 和 AMP 模式：

![](https://user-images.githubusercontent.com/35585791/211782053-b1e3d7a7-ff12-44ba-b73f-cc55231ae662.png)

可以发现将原始 GLM 的 naive sparse_softmax_cross_entropy 实现替换为 `flow.sparse_softmax_cross_entropy` 之后 loss 是可以完全对齐的，可以保证正确性。

这个替换使得 one-glm 的单卡性能相比于原始的 GLM 的单卡性能有大幅度提升，主要原因就是因为 OneFlow 对 `sparse_softmax_cross_entropy` 算子做了极致的性能优化并且减少了原始 GLM 中大量的碎算子拼凑带来的访存开销。此外，也降低了 `torch.autograd.Function` 本身带来的一些系统开销。

### cuda kernel fuse

除上述优化外，我们发现 GLM 模型本质上就是一个编解码的 Transformer 架构。我们将之前优化 GPT，BERT 的一些 Fuse Pattern 也带给了 one-glm 模型。具体包含以下两个 Fuse Pattern :

- [fused_bias_add_gelu](https://github.com/Oneflow-Inc/one-glm/blob/fe38460185fc5a232a3effd77ca242c5e19ecd57/mpu/layers.py#L265): 将 bias_add 和 gelu 算子融合在一起。
- [fused_bias_add_dropout](https://github.com/Oneflow-Inc/one-glm/blob/fe38460185fc5a232a3effd77ca242c5e19ecd57/mpu/layers.py#L269)：将 bias_add 和 dropout 算子融合在一起。

这两个 fuse 都可以显著改善计算的访存并且减少 Kernel Launch 带来的开销，由于 GLM 模型越大则层数就会越多，那么这种 Fuse Pattern 带来的的优势也会不断放大。

最终，在上面两方面的优化作用下，我们在 A100 PCIE 40G 并且 batch_size = 32 的训练 GLM-large 的任务中，单卡下的性能相比于原始的 GLM 取得了 3.1（FP32模式） 倍和 1.65（AMP模式） 倍的训练加速。

## 用LiBai搭建GLM的推理

接下来这部分我们将介绍如何利用基于OneFlow构建的大规模模型训练与推理仓库LiBai来构建GLM的分布式推理部分。
- LiBai链接：https://github.com/Oneflow-Inc/libai
- GLM项目链接：https://github.com/Oneflow-Inc/libai/tree/glm_project/projects/GLM

### 实现分布式推理的细节原理
用LiBai搭建的GLM可以便捷的实现`model parallel + pipeline parallel`推理, 较好的解决单卡放不下大规模模型的问题，我们用下面的例子说明为什么libai可以较好的实现模型的分布式推理：
- 模型的参数其实就是许多`tensor`，也就是以矩阵的形式出现，大模型的参数也就是大矩阵，并行策略就是把大矩阵分为多个小矩阵分配到不同的显卡或者不同的设备上，基础的[LinearLayer](https://github.com/Oneflow-Inc/libai/blob/main/libai/layers/linear.py)在`LiBai`中的实现代码如下。在这里用户可选择去如何切`Linear`层的矩阵（模型并行），如何切分数据矩阵（数据并行），`OneFlow`中的`sbp` 也就是控制是竖着切、横着切以及其他拆分矩阵的方案，以及通过设置`placement`来控制这个`LinearLayer`是放在第几张显卡上（流水并行）。
```python
class Linear1D(nn.Module):
    def __init__(self, in_features, out_features, parallel="data", layer_idx=0, ...):
        super().__init__()

        if parallel == "col":
            weight_sbp = dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.split(0)])
        elif parallel == "row":
            weight_sbp = dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.split(1)])
        elif parallel == "data":
            weight_sbp = dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast])
        else:
            raise KeyError(f"{parallel} is not supported! Only support ('data', 'row' and 'col')")

        self.weight = flow.nn.Parameter(
            flow.empty(
                (out_features, in_features),
                dtype=flow.float32,
                placement=dist.get_layer_placement(layer_idx),  # for pipeline parallelism placement
                sbp=weight_sbp,
            )
        )
        init_method(self.weight)
        ...
    
    def forward(self, x):
        ...
```
所以根据`LiBai`中各种`layer`的设计原理以及基于`OneFlow`中`tensor`自带的`sbp`和`placement`属性的天然优势，搭建出来的模型能够简单的实现数据并行、模型并行以及流水并行，感兴趣的用户可以查看更多的矩阵切分[原理和方案](https://docs.oneflow.org/master/parallelism/02_sbp.html#sbp)。

### GLM推理demo演示
下面我们展示libai种GLM的单卡和多卡推理demo：
- 单卡generate任务：
```bash
python demo.py
```
```python
# demo.py
import oneflow as flow
from projects.GLM.tokenizer.glm_tokenizer import GLMGPT2Tokenizer
from libai.utils import distributed as dist
from projects.GLM.configs.glm_inference import cfg
from projects.GLM.modeling_glm import GLMForConditionalGeneration
from projects.GLM.utils.glm_loader import GLMLoaderHuggerFace
from omegaconf import DictConfig

tokenizer = GLMGPT2Tokenizer.from_pretrained("/data/home/xiezipeng/glm-10b")
input_ids = tokenizer.encode(
    [
        "Ng is an adjunct professor at [MASK] (formerly associate professor and Director of its Stanford AI Lab or SAIL ). Also a pioneer in online education, Ng co-founded Coursera and deeplearning.ai."
    ],
    return_tensors="of",
)
inputs = {"input_ids": input_ids, "attention_mask": flow.ones(input_ids.size())}
inputs = tokenizer.build_inputs_for_generation(inputs, max_gen_length=512)

sbp = dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast])
placement = dist.get_layer_placement(0)

dist.set_device_type("cpu")
loader = GLMLoaderHuggerFace(GLMForConditionalGeneration, cfg, "/path/to/glm-10b")
model = loader.load()
model = model.half().cuda()

dist.set_device_type("cuda")
outputs = model.generate(
    inputs=inputs['input_ids'].to_global(sbp=sbp, placement=placement), 
    position_ids=inputs['position_ids'].to_global(sbp=sbp, placement=placement), 
    generation_attention_mask=inputs['generation_attention_mask'].to_global(sbp=sbp, placement=placement).half(), 
    max_length=512
)
res = tokenizer.decode(outputs[0])
print(res)
>>> [CLS] Ng is an adjunct professor at [MASK] (formerly associate professor and Director of its Stanford AI Lab or SAIL ). Also a pioneer in online education, Ng co-founded Coursera and deeplearning.ai.<|endoftext|> <|startofpiece|>  Stanford University and a co-founder of <|endofpiece|>
```

- 4卡model parallel+pipeline parallel generate任务
```bash
python3 -m oneflow.distributed.launch --nproc_per_node 4 demo.py 
```
```python
# demo.py
import oneflow as flow
from projects.GLM.tokenizer.glm_tokenizer import GLMGPT2Tokenizer
from libai.utils import distributed as dist
from projects.GLM.configs.glm_inference import cfg
from projects.GLM.modeling_glm import GLMForConditionalGeneration
from projects.GLM.utils.glm_loader import GLMLoaderHuggerFace
from omegaconf import DictConfig

parallel_config = DictConfig(
    dict(
        data_parallel_size=1,
        tensor_parallel_size=2,
        pipeline_parallel_size=2,
        pipeline_num_layers=2 * 24
    )
)
dist.setup_dist_util(parallel_config)

tokenizer = GLMGPT2Tokenizer.from_pretrained("/data/home/xiezipeng/glm-10b")
input_ids = tokenizer.encode(
    [
        "Ng is an adjunct professor at [MASK] (formerly associate professor and Director of its Stanford AI Lab or SAIL ). Also a pioneer in online education, Ng co-founded Coursera and deeplearning.ai."
    ],
    return_tensors="of",
)
inputs = {"input_ids": input_ids, "attention_mask": flow.ones(input_ids.size())}
inputs = tokenizer.build_inputs_for_generation(inputs, max_gen_length=512)

sbp = dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast])
placement = dist.get_layer_placement(0)

loader = GLMLoaderHuggerFace(GLMForConditionalGeneration, cfg, "/path/to/glm-10b")
model = loader.load()

outputs = model.generate(
    inputs=inputs['input_ids'].to_global(sbp=sbp, placement=placement), 
    position_ids=inputs['position_ids'].to_global(sbp=sbp, placement=placement), 
    generation_attention_mask=inputs['generation_attention_mask'].to_global(sbp=sbp, placement=placement), 
    max_length=512
)
res = tokenizer.decode(outputs[0])
if dist.is_main_process():
    print(res)
>>> [CLS] Ng is an adjunct professor at [MASK] (formerly associate professor and Director of its Stanford AI Lab or SAIL ). Also a pioneer in online education, Ng co-founded Coursera and deeplearning.ai.<|endoftext|> <|startofpiece|>  Stanford University and a co-founder of <|endofpiece|>
```

## 总结

本文介绍了使用 OneFlow 深度学习框架移植 GLM 并做简单调优之后相比于原始 GLM 做预训练任务可以取得的大幅性能以及显存优势。我们相信在大模型训练方面，OneFlow的易用性和性能都是经得起实际检验的。

**Comment by whalefa1I** [comment]
您好，请问GLM的预训练数据处理样例方便提供一下吗，想用oneflow复现一下但是没有/glm_train.txt

**Comment by erichen510** [comment]
module 'oneflow._C' has no attribute 'fused_multi_head_attention_inference_v2' oneflow好像找不到这个kernel
